include/onnxruntime/core/optimizer/graph_transformer.h

```diff
@@ -36,7 +42,7 @@ class GraphTransformer {
   @param[out] modified Set to true if the Graph was modified.
   @returns Status with success or error information.
   */
-  common::Status Apply(Graph& graph, bool& modified) const;
+  common::Status Apply(Graph& graph, bool& modified, const std::vector<std::string>& providers) const;  
 
```

Comment:
Defaulting providers to {} would make some of the calling code a bit cleaner. Or have two Apply methods so you can document that the one without providers is 'global'.

  common::Status Apply(Graph& graph, bool& modified, const std::vector<std::string>& providers) const;  
  common::Status Apply(Graph& graph, bool& modified) const { return Apply(graph, modified, {}); }

Reply:
Is the providers the target providers or the providers that this transformation supports?
I feel we are adding too much customization capabilities here without being sure that we need them yet.
Say I have enabled L2 and that should run only for specific execution providers. I will not even register transformations for other providers right?
If I am not missing something, this will avoid a lot of code and complexity below.

Reply:
So I am thinking if we even need a providers list here.
Unless we eventually want to get to having transformers look at the exec-provider a specific node has been assigned to and act accordingly. But I would say to get first a version that does not account for such cases.

Reply:
We do need providers here. A transformer needs to know which providers it can operate on. If it is ep agnostic this list will be empty.
We already have transformers which will need this list. Example ConvActivationFusion needs to know which EPs it can work with. In current scenario it is only CPU. So in the case when Conv is assigned to CUDA , this transformer will not fuse Conv and Activation

Reply:
What I am still not sure about is how this will work in the case of rewrite-rules.
You will have rewrite rules that will be compatible with some execution providers and some others that will be compatible with others. Will such rules be registered to the same rule-based transformer?

Reply:
Right now we dont have such a case... but yes if we do have different rules which are not compatible with same providers then we should bind them to different transformers... But again we dont have such a case right now... In fact this what we discussed offline too

Reply:
Let's leave it as is for now -- we revisit once we actually implement the transformers with exec providers.